Avoid overriding CUDA_ARCH if CMAKE_CUDA_ARCHITECTURES is non-empty
diff -Nru tsne-cuda-3.0.1.orig/CMakeLists.txt tsne-cuda-3.0.1/CMakeLists.txt
--- tsne-cuda-3.0.1.orig/CMakeLists.txt	2024-01-18 17:31:58.841767793 +0000
+++ tsne-cuda-3.0.1/CMakeLists.txt	2024-01-18 18:17:35.153863840 +0000
@@ -58,55 +58,56 @@
     set(CMAKE_CUDA_STANDARD_REQUIRED ON)
 endif()
 
-if(CUDAToolkit_VERSION_MAJOR EQUAL "10")
-    set(CUDA_ARCH
-        -gencode=arch=compute_30,code=sm_30
-        -gencode=arch=compute_35,code=sm_35
-        -gencode=arch=compute_37,code=sm_37
-        -gencode=arch=compute_50,code=sm_50
-        -gencode=arch=compute_52,code=sm_52
-        -gencode=arch=compute_60,code=sm_60
-        -gencode=arch=compute_61,code=sm_61
-        -gencode=arch=compute_70,code=sm_70
-        -gencode=arch=compute_75,code=sm_75
-    )
-elseif(CUDAToolkit_VERSION_MAJOR EQUAL "11" AND CUDAToolkit_VERSION_MINOR LESS "1")
-    set(CUDA_ARCH
-        -gencode=arch=compute_35,code=sm_35
-        -gencode=arch=compute_37,code=sm_37
-        -gencode=arch=compute_50,code=sm_50
-        -gencode=arch=compute_52,code=sm_52
-        -gencode=arch=compute_60,code=sm_60
-        -gencode=arch=compute_61,code=sm_61
-        -gencode=arch=compute_70,code=sm_70
-        -gencode=arch=compute_75,code=sm_75
-        -gencode=arch=compute_80,code=sm_80
-    )
-elseif(CUDAToolkit_VERSION_MAJOR EQUAL "11")
-    set(CUDA_ARCH
-        -gencode=arch=compute_35,code=sm_35
-        -gencode=arch=compute_37,code=sm_37
-        -gencode=arch=compute_50,code=sm_50
-        -gencode=arch=compute_52,code=sm_52
-        -gencode=arch=compute_60,code=sm_60
-        -gencode=arch=compute_61,code=sm_61
-        -gencode=arch=compute_70,code=sm_70
-        -gencode=arch=compute_75,code=sm_75
-        -gencode=arch=compute_80,code=sm_80
-        -gencode=arch=compute_86,code=sm_86
-    )
-else()
-    set(CUDA_ARCH
-        -gencode=arch=compute_30,code=sm_30
-        -gencode=arch=compute_35,code=sm_35
-        -gencode=arch=compute_37,code=sm_37
-        -gencode=arch=compute_50,code=sm_50
-        -gencode=arch=compute_52,code=sm_52
-        -gencode=arch=compute_60,code=sm_60
-        -gencode=arch=compute_61,code=sm_61
+if("${CMAKE_CUDA_ARCHITECTURES}" STREQUAL "")
+    if(CUDAToolkit_VERSION_MAJOR EQUAL "10")
+        set(CUDA_ARCH
+            -gencode=arch=compute_30,code=sm_30
+            -gencode=arch=compute_35,code=sm_35
+            -gencode=arch=compute_37,code=sm_37
+            -gencode=arch=compute_50,code=sm_50
+            -gencode=arch=compute_52,code=sm_52
+            -gencode=arch=compute_60,code=sm_60
+            -gencode=arch=compute_61,code=sm_61
+            -gencode=arch=compute_70,code=sm_70
+            -gencode=arch=compute_75,code=sm_75
         )
-endif()
-
+    elseif(CUDAToolkit_VERSION_MAJOR EQUAL "11" AND CUDAToolkit_VERSION_MINOR LESS "1")
+        set(CUDA_ARCH
+            -gencode=arch=compute_35,code=sm_35
+            -gencode=arch=compute_37,code=sm_37
+            -gencode=arch=compute_50,code=sm_50
+            -gencode=arch=compute_52,code=sm_52
+            -gencode=arch=compute_60,code=sm_60
+            -gencode=arch=compute_61,code=sm_61
+            -gencode=arch=compute_70,code=sm_70
+            -gencode=arch=compute_75,code=sm_75
+            -gencode=arch=compute_80,code=sm_80
+        )
+    elseif(CUDAToolkit_VERSION_MAJOR EQUAL "11")
+        set(CUDA_ARCH
+            -gencode=arch=compute_35,code=sm_35
+            -gencode=arch=compute_37,code=sm_37
+            -gencode=arch=compute_50,code=sm_50
+            -gencode=arch=compute_52,code=sm_52
+            -gencode=arch=compute_60,code=sm_60
+            -gencode=arch=compute_61,code=sm_61
+            -gencode=arch=compute_70,code=sm_70
+            -gencode=arch=compute_75,code=sm_75
+            -gencode=arch=compute_80,code=sm_80
+            -gencode=arch=compute_86,code=sm_86
+        )
+    else()
+        set(CUDA_ARCH
+            -gencode=arch=compute_30,code=sm_30
+            -gencode=arch=compute_35,code=sm_35
+            -gencode=arch=compute_37,code=sm_37
+            -gencode=arch=compute_50,code=sm_50
+            -gencode=arch=compute_52,code=sm_52
+            -gencode=arch=compute_60,code=sm_60
+            -gencode=arch=compute_61,code=sm_61
+            )
+    endif()
+endif()
 
 set(CUDA_OPTS
     -O3
